ramips: fix PCIe NIC initialization issues on ZBT-WE1326

DragonBluep opened the pull request:
Some users report that their Wi-Fi interface is lost after reboot, which may be caused by allocating the wrong pcie port. This commit fixes these wrong ports.

For Netgear R6220, WAC104 and WNDR3700 v5, according to bootlog, MT7612E (5GHz) is connected to pcie0, and MT7603E (2GHz) is connected to pcie2:
```
[2.758986] mt7621-pci 1e140000.pcie: pcie1 no card, disable it (RST & CLK)
[2.772862] mt7621-pci 1e140000.pcie: PCIE0 enabled
[2.782579] mt7621-pci 1e140000.pcie: PCIE2 enabled
...
[3.009151] pci 0000:01:00.0: [14c3:7662] type 00 class 0x028000
[3.125715] pci 0000:02:00.0: [14c3:7603] type 00 class 0x028000
```
For Zbtlink ZBT-WE1326, according to bootlog, MT7612E (5GHz) is connected to pcie1, and MT7603E (2GHz) is connected to pcie2:
```
[4.197658] mt7621-pci 1e140000.pcie: pcie0 no card, disable it (RST & CLK)
[4.204609] mt7621-pci 1e140000.pcie: PCIE1 enabled
[4.209476] mt7621-pci 1e140000.pcie: PCIE2 enabled
...
[4.307988] pci 0000:01:00.0: [14c3:7662] type 00 class 0x028000
[4.367206] pci 0000:02:00.0: [14c3:7603] type 00 class 0x028000
```
OpenWrt Forum Link:
[Missing 2.4GHz interface on WAC104 & difficulty setting up forwarding](https://forum.openwrt.org/t/missing-2-4ghz-interface-on-wac104-difficulty-setting-up-forwarding/119855)
[R6220 ” Could not find PHY for device ‘radio0’” mt76x2 .. issue continues](https://forum.openwrt.org/t/r6220-could-not-find-phy-for-device-radio0-mt76x2-issue-continues/116665)
[ZBT-1326 OpenWRT stable 21.02 results in no Wifi](https://forum.openwrt.org/t/zbt-1326-openwrt-stable-21-02-results-in-no-wifi/108116)

GitHub issues:
https://github.com/openwrt/mt76/issues/199
https://github.com/openwrt/mt76/issues/601
https://github.com/openwrt/openwrt/issues/9374

**I have no relevant device to test, but this should be the correct fix.**

DragonBluep commented:
Test images for R6220, WAC104, WNDR3700 v5 and ZBT-WE1326:
https://we.tl/t-jplW7zU4CX

wifiseguro commented:
I have just tried this on my zbt we1326 and unfortunatelly these doesn't solve the issue I am having since release 19.

Have compiled with the changes done on the dts file, flashed and no 2.4 wifi. 
LOG:

Sat Sep 10 11:03:30 2022 kern.err kernel: [    1.288233] mt7621-pci 1e140000.pcie: pcie0 no card, disable it (RST & CLK)
Sat Sep 10 11:03:30 2022 kern.err kernel: [    1.295165] mt7621-pci 1e140000.pcie: pcie2 no card, disable it (RST & CLK)
Sat Sep 10 11:03:30 2022 kern.info kernel: [    1.302113] mt7621-pci 1e140000.pcie: PCIE1 enabled

LUCI:

![Captura desde 2022-11-12 20-41-44](https://user-images.githubusercontent.com/46748343/201493108-05c4f325-5fcd-4286-a48b-66d901a21bc0.png)

DragonBluep commented:
@wifiseguro Can you try the latest patch?
Reference: https://github.com/openwrt/openwrt/commit/6f2c95f0cf0519631adb5100ab71f908a2968cca https://github.com/openwrt/openwrt/commit/06356f00200639c48d95330e633965957b0347ab https://github.com/openwrt/openwrt/commit/af1b6799c6ec9af7a30d63a5ddfed20f443b991c

DragonBluep commented:
Test images for R6220, WAC104, WNDR3700 v5 and ZBT-WE1326:
https://we.tl/t-SpWefIk3ib

wifiseguro commented:
> @wifiseguro Can you try the latest patch? Reference: [6f2c95f](https://github.com/openwrt/openwrt/commit/6f2c95f0cf0519631adb5100ab71f908a2968cca) [06356f0](https://github.com/openwrt/openwrt/commit/06356f00200639c48d95330e633965957b0347ab) [af1b679](https://github.com/openwrt/openwrt/commit/af1b6799c6ec9af7a30d63a5ddfed20f443b991c) And #9810 ramips: zbt-wg2626: Add the reset gpio for PCIe port 1

@DragonBluep  Thank you again for all your efforts for having this device working properly.
Unfortunatelly still have had no luck with my zbt we1326.
I tried your patchs on 22.03 cloning todays souce and complilling. These are the results:
With just the Gpios reset part on the dts file, nothing changes, still no 2.4 wifi band.
If I compille with both the gpios reset and pci numbering corrected (pcie1 and pcie2 instead of pcie0 and pcie1) the router does not reboot and I have to enter vendors safemode to flash a new firmware (pinning reset button while powering on)
I noticed this boot issue yesterday when tried the renunbering of the pcie ports, I mean that it occurs even when applying only this part of the patch without the gpios reset.
I look forward your comments.

DragonBluep commented:
Does your router have a 3G/4G module or other pcie device?  I see a pcie slot on the board.
The bootlog I found on the forum shows `pcie0 no card, disable it (RST&CLK)`. So I'm pretty sure we only use pcie1 and pcie2.

DragonBluep commented:
Test Firmware: https://we.tl/t-8MWLM0P0W0

DragonBluep commented:
@paraka Can you provide some suggestions? Great thanks!

wifiseguro commented:
> So normally when the detection fails while the PCI bus is being set up is because one of this two problems:
> 
> * One GPIO has not properly been reset.
**> * Time after reseting stuff is not enough.**
> 
> If the problem is related with timing, the way to go should be to change **PERS_DELAY_MS definition [0]**(https://elixir.bootlin.com/linux/v6.1-rc1/source/drivers/pci/controller/pcie-mt7621.c#L63) since this is the one used after the related resets [1](https://elixir.bootlin.com/linux/v6.1-rc1/source/drivers/pci/controller/pcie-mt7621.c#L326) and [2](https://elixir.bootlin.com/linux/v6.1-rc1/source/drivers/pci/controller/pcie-mt7621.c#L344).

What I tried on a  ZBT WE1326:
Increased **PERS_DELAY_MS definition [0]** to 200 ms -> Issue is still there.
Increased **PERS_DELAY_MS definition [0]** to 400 ms -> Issue Fixed :-)
I belive this might help other devices, not just ZBT WE1326, because applying the DELAY in a similar way (https://github.com/openwrt/openwrt/pull/10985) also fixes the pcie iniatialization for the Netgear R6220
Of course, Paraka's way of applying the delay is the proper way. I will mention there just in case can be changed and done properly.

dfateyev commented:
Could anybody clarify whether we need DTS changes for the devices (proposed in this PR) if we introduce [PERST_DELAY_MS](https://elixir.bootlin.com/linux/v6.1-rc1/source/drivers/pci/controller/pcie-mt7621.c#L63) value increase?

paraka commented:
> 
> What I tried on a ZBT WE1326: Increased **PERS_DELAY_MS definition [0]** to 200 ms -> Issue is still there. Increased **PERS_DELAY_MS definition [0]** to 400 ms -> Issue Fixed :-) 

Thanks for letting me know. I have just sent a patch increasing this value to 500 ms to the mainline [0]. Hope this helps!

[0]: https://lore.kernel.org/linux-pci/20221119110837.2419466-1-sergio.paracuellos@gmail.com/T/#u

paraka commented:
> Could anybody clarify whether we need DTS changes for the devices (proposed in this PR) if we introduce [PERST_DELAY_MS](https://elixir.bootlin.com/linux/v6.1-rc1/source/drivers/pci/controller/pcie-mt7621.c#L63) value increase?

DTS changes are not related with pcie port detection but for the correct way interrupts are expanded along the PCI bus once are properly detected since new driver in use now use _interrupt-maps_ and _interrupt-map-mask properties_ in general mt7621.dtsi file. If your problem is with initial detection only bad gpio reset or a timing issue applies, AFAICT.

wifiseguro commented:
> > What I tried on a ZBT WE1326: Increased **PERS_DELAY_MS definition [0]** to 200 ms -> Issue is still there. Increased **PERS_DELAY_MS definition [0]** to 400 ms -> Issue Fixed :-)
> 
> Thanks for letting me know. I have just sent a patch increasing this value to 500 ms to the mainline [0]. Hope this helps!
> 
> [0]: https://lore.kernel.org/linux-pci/20221119110837.2419466-1-sergio.paracuellos@gmail.com/T/#u

First of all, thanks for all your work and explanations, it really means a lot to me
Today Ive been doing a lot of tests to be sure we are sending the right patch, dont wanna rush to have it in the mainline if I am not sure that we really found a solution, after all, I've been suffering this issue for many yeas(I have been applying a patch to my own builds that worked flawless though), but having a fix in the mainline is far more satisfacing and many other can benefit of it.
Well, lets explain the tests I've done today, hoping you can help to figure out:

1 - I first tried a build with  a 400ms delay, then 500 and finally 750 ms patch and the issue was gone. So why I was trying longer delays if 400ms was enough? The reason is weird: when I warm reboot the device using Luci o reboot command, the issue is back. If I cold reboot, removing the power adaptor and waiting a few seconds there is no issue detecting both pcie cards. Weird isn't it?
After doing many cold and warm reboots to be sure that this behaviour is repetitive, I decided to try again my previous patch, which has been working many years with no issue, no matter I warm or cold reboot.
So I restore  the PERS_DELAY_MS to 100ms and I decide to dig a bit on my own patch, and I dont understand why, but the conclusion, also after many reboots with this patch, was that my device needs just 75ms delay!, but placed in a diferent place of the firmware code. 
I am putting just 75 ms delay right before function mt7621_pcie_reset_ep_deassert(pcie) and that fixes 2.4 pcie card detection, as I told, in this case no matter I cold or warm reboot the device.
My previous version of the patch was applying a 250 ms delay, not only where I mentioned but also after the resets, just like the PERS_DELAY_MS does. Though definitelly are not necessary in my device, and dont make any diferrence.
Hope this is worth to send a patch, at least to be applied to the zbt we1326 device.

This is how the patch looks like:

--- a/drivers/staging/mt7621-pci/pci-mt7621.c
+++ b/drivers/staging/mt7621-pci/pci-mt7621.c
@@ -520,6 +520,7 @@
                        list_del(&port->list);
                }
        }
+       mdelay(75);

        mt7621_pcie_reset_ep_deassert(pcie);

I hope I have explained this case properly (my native languaje is spanish)
We look forward your comments.

wifiseguro commented:
> Could anybody clarify whether we need DTS changes for the devices (proposed in this PR) if we introduce [PERST_DELAY_MS](https://elixir.bootlin.com/linux/v6.1-rc1/source/drivers/pci/controller/pcie-mt7621.c#L63) value increase?

At least, on my zbt we1326 device is not needed. It just soft briks the device.

paraka commented:
> 

@wifiseguro Does changing mdelay with msleep and putting inside mt7621_pcie_reset_rc_deassert function works for you in both warn and cold resets? Something like the following:

```
@@ -332,6 +332,8 @@ static void mt7621_pcie_reset_rc_deassert(struct mt7621_pcie *pcie)
 
        list_for_each_entry(port, &pcie->ports, list)
                mt7621_control_deassert(port);
+
+       msleep(PERST_DELAY_MS);
 }
 
 static void mt7621_pcie_reset_ep_deassert(struct mt7621_pcie *pcie)
```

wifiseguro commented:
> > 
> 
> @wifiseguro Does changing mdelay with msleep and putting inside mt7621_pcie_reset_rc_deassert function works for you in both warn and cold resets? Something like the following:
> 
> ```
> @@ -332,6 +332,8 @@ static void mt7621_pcie_reset_rc_deassert(struct mt7621_pcie *pcie)
>  
>         list_for_each_entry(port, &pcie->ports, list)
>                 mt7621_control_deassert(port);
> +
> +       msleep(PERST_DELAY_MS);
>  }
>  
>  static void mt7621_pcie_reset_ep_deassert(struct mt7621_pcie *pcie)
> ```

I have just build with that patch. Results:
On first boot right after flashing: Both radios came up.
When I reboot, tried several times cold and warm, just one radio come up.
There is a detail I want to mention:
When no patch is applied, no msdelay, no msleep, Luci shows only one card, MT76x2E 802.11bgnac. 
And when I apply one of the differents patches,  on warm reboot, Luci shows a Generic bgn card, instead of MT7603E.
Despite logs show same error in both cases:

> Sat Sep 10 11:03:30 2022 kern.err kernel: [ 1.288233] mt7621-pci 1e140000.pcie: pcie0 no card, disable it (RST & CLK)
> Sat Sep 10 11:03:30 2022 kern.err kernel: [ 1.295165] mt7621-pci 1e140000.pcie: pcie2 no card, disable it (RST & CLK)
> Sat Sep 10 11:03:30 2022 kern.info kernel: [ 1.302113] mt7621-pci 1e140000.pcie: PCIE1 enabled

Luci Generic bg 
![Captura desde 2022-11-20 11-31-47](https://user-images.githubusercontent.com/46748343/202897782-cd7a0cab-0176-4eda-ba17-f42ad1d4ade4.png)

Luci no card at all
![Captura desde 2022-11-20 11-39-59](https://user-images.githubusercontent.com/46748343/202897826-a1d83ad2-ea3f-4c76-b97c-ef2afe6b9176.png)

DragonBluep commented:
@wifiseguro Can you test again with the second commit in this PR `ramips: overwrite reset gpio properties in ZBT WE1326` and your `delay patch`.

paraka commented:
@wifiseguro Weird... It looks like your HW needs to wait a bit after making the phy_power-on of the physical port... Does your patch work if you replace mdelay with msleep?

wifiseguro commented:
> @wifiseguro Can you test again with the second commit in this PR `ramips: overwrite reset gpio properties in ZBT WE1326` and your `delay patch`.

Built as suggested, with second commit "overwrite reset gpio properties" and  my delay patch (75ms delay before mt7621_pcie_reset_ep_deassert(pcie) function.)
Result:
Device working properly no issue of any kind. And I can warm or cold reboot with no issues either.
I Don't know what the "reset gpio" does but for sure my device is a little better now.
Thank you!

wifiseguro commented:
> @wifiseguro Weird... It looks like your HW needs to wait a bit after making the phy_power-on of the physical port... Does your patch work if you replace mdelay with msleep?
Built with msleep and the patch works. All issues are gone with msleep too. :-)

wifiseguro commented:
> > Could anybody clarify whether we need DTS changes for the devices (proposed in this PR) if we introduce [PERST_DELAY_MS](https://elixir.bootlin.com/linux/v6.1-rc1/source/drivers/pci/controller/pcie-mt7621.c#L63) value increase?
> 
> At least, on my zbt we1326 device is not needed. It just soft briks the device.

@DragonBluep @dfateyev
Are thre any recent changes in the DTS update to correct the pcie port numberfor zbt we1326?
I wonder becaise I have just built and now it does boot, but it doesn't detect any card, consequentelly LUCI shows no wireless menu.
>Sun Nov 13 17:11:41 2022 kern.err kernel: [    1.367603] mt7621-pci 1e140000.pcie: pcie1 no card, disable it (RST & CLK)
>Sun Nov 13 17:11:41 2022 kern.err kernel: [    1.374554] mt7621-pci 1e140000.pcie: pcie2 no card, disable it (RST & CLK)
>Sun Nov 13 17:11:41 2022 kern.err kernel: [    1.381500] mt7621-pci 1e140000.pcie: Nothing is connected in virtual bridges. Exiting...
Note I have built with both commits and the msleep(75) pacth

DragonBluep commented:
@wifiseguro This change will only take effect on 5.15 kernel.

> See [0](https://www.spinics.net/lists/linux-driver-devel/msg133251.html) where old legacy driver approach was used because the bad understanding of the renamed stuff. And finally [1](https://www.spinics.net/lists/linux-staging/msg02512.html) and [2](https://www.spinics.net/lists/linux-staging/msg02515.html) where the properties are used and interrupts are properly mapped along the buses as they are without any kind of internal register configuration.

It seems that `interrupt-map` have been configured correctly. I'm still not clear about it.
https://github.com/openwrt/openwrt/blob/0d375de10dac3160c65c264bb91a5137ef4c0817/target/linux/ramips/dts/mt7621.dtsi#L576-L671

paraka commented:
> > @wifiseguro Weird... It looks like your HW needs to wait a bit after making the phy_power-on of the physical port... Does your patch work if you replace mdelay with msleep?
> 
> @paraka Built with msleep and the patch works. All issues are gone with msleep too. :-) Would it be "msleep" a more proper way to fix it?

@wifiseguro Kernel maintainers prefer to use msleep where possible when patches are submited, so I needed to know.

@DragonBluep  BTW, Does the netgear router also affected with similar issue work with this msleep or PERS_DELAY value needs also to be set to 250???

DragonBluep commented:
> BTW, Does the netgear router also affected with similar issue work with this msleep or PERS_DELAY value needs also to be set to 250???

I don't have relevant device to test, so I'm not sure. I just found that they used the wrong pcie port, so I opened this PR.
From another PR https://github.com/openwrt/openwrt/pull/10985 I think maybe yes.
Although until now I can't figure out why they need such a long delay. 100ms is long enough for these high-speed pcie devices. A reasonable explanation may be that some gpio control the power supply of the pcie, but the u-boot does not initialize them.

dfateyev commented:
Zbtlink ZBT WE3526: I faced the same issues described by @wifiseguro above.
The MT7603E 2.4GHz interface is not detected in most attempts on 5.15 after a cold or warm reboot. The same situation was on 5.10 with 22.03 builds, where I was applying a [patch](https://github.com/openwrt/openwrt/issues/9374#issuecomment-1244150080) to get it working.
What I have tried:
- `PERST_DELAY_MS = 500`: decreased initialization fails by half, but not excluded;
- `PERST_DELAY_MS = 100` and `msleep(PERST_DELAY_MS)`: slightly worse than the option above;
- `PERST_DELAY_MS = 500` and `msleep(PERST_DELAY_MS)`: slightly better than the option above, but detection fails still present.

```
BusyBox v1.35.0 (2022-11-20 17:44:43 UTC) built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt SNAPSHOT, r21333-0d375de10d
 -----------------------------------------------------
root@OpenWrt:~# dmesg
[    0.000000] Linux version 5.15.79 (builder@builder) (mipsel-openwrt-linux-musl-gcc (OpenWrt GCC 11.3.0 r21335-e410833bdd) 11.3.0, GNU ld (GNU Binutils) 2.37) #0 SMP Sun Nov 20 17:44:43 2022
[    0.000000] SoC Type: MediaTek MT7621 ver:1 eco:3
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0001992f (MIPS 1004Kc)
[    0.000000] MIPS: machine is Zbtlink ZBT-WE3526
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] VPE topology {2,2} total 4
[    0.000000] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x000000000fffffff]
[    0.000000]   HighMem  empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x000000000fffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x000000000fffffff]
[    0.000000] percpu: Embedded 11 pages/cpu s15632 r8192 d21232 u45056
[    0.000000] pcpu-alloc: s15632 r8192 d21232 u45056 alloc=11*4096
[    0.000000] pcpu-alloc: [0] 0 [0] 1 [0] 2 [0] 3 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 64960
[    0.000000] Kernel command line: console=ttyS0,115200 rootfstype=squashfs,jffs2
[    0.000000] Dentry cache hash table entries: 32768 (order: 5, 131072 bytes, linear)
[    0.000000] Inode-cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Writing ErrCtl register=00043340
[    0.000000] Readback ErrCtl register=00043340
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 248716K/262144K available (7125K kernel code, 625K rwdata, 1444K rodata, 1224K init, 242K bss, 13428K reserved, 0K cma-reserved, 0K highmem)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=4, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000] 	Tracing variant of Tasks RCU enabled.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 10 jiffies.
[    0.000000] NR_IRQS: 256
[    0.000000] clocksource: GIC: mask: 0xffffffffffffffff max_cycles: 0xcaf478abb4, max_idle_ns: 440795247997 ns
[    0.000004] sched_clock: 64 bits at 880MHz, resolution 1ns, wraps every 4398046511103ns
[    0.008074] Calibrating delay loop... 586.13 BogoMIPS (lpj=2930688)
[    0.066214] pid_max: default: 32768 minimum: 301
[    0.071036] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.078248] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.089291] rcu: Hierarchical SRCU implementation.
[    0.094346] dyndbg: Ignore empty _ddebug table in a CONFIG_DYNAMIC_DEBUG_CORE build
[    0.102490] smp: Bringing up secondary CPUs ...
[    0.107800] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.107824] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.107839] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.107887] CPU1 revision is: 0001992f (MIPS 1004Kc)
[    0.162254] Synchronize counters for CPU 1: done.
[    0.194558] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.194580] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.194591] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.194619] CPU2 revision is: 0001992f (MIPS 1004Kc)
[    0.253662] Synchronize counters for CPU 2: done.
[    0.284294] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.284315] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.284326] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.284357] CPU3 revision is: 0001992f (MIPS 1004Kc)
[    0.338859] Synchronize counters for CPU 3: done.
[    0.368725] smp: Brought up 1 node, 4 CPUs
[    0.376854] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.386663] futex hash table entries: 1024 (order: 3, 32768 bytes, linear)
[    0.393701] pinctrl core: initialized pinctrl subsystem
[    0.400215] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.407158] cpuidle: using governor teo
[    0.422471] FPU Affinity set after 11720 emulations
[    0.434549] clocksource: Switched to clocksource GIC
[    0.440335] NET: Registered PF_INET protocol family
[    0.445368] IP idents hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.453232] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 6144 bytes, linear)
[    0.461568] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.469238] TCP established hash table entries: 2048 (order: 1, 8192 bytes, linear)
[    0.476866] TCP bind hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.483934] TCP: Hash tables configured (established 2048 bind 2048)
[    0.490392] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.496891] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.504094] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.509729] PCI: CLS 0 bytes, default 32
[    0.516420] workingset: timestamp_bits=14 max_order=16 bucket_order=2
[    0.527671] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.533433] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.547880] mt7621_gpio 1e000600.gpio: registering 32 gpios
[    0.553816] mt7621_gpio 1e000600.gpio: registering 32 gpios
[    0.559797] mt7621_gpio 1e000600.gpio: registering 32 gpios
[    0.565986] mt7621-pci 1e140000.pcie: host bridge /pcie@1e140000 ranges:
[    0.572651] mt7621-pci 1e140000.pcie:   No bus range found for /pcie@1e140000, using [bus 00-ff]
[    0.581467] mt7621-pci 1e140000.pcie:      MEM 0x0060000000..0x006fffffff -> 0x0060000000
[    0.589580] mt7621-pci 1e140000.pcie:       IO 0x001e160000..0x001e16ffff -> 0x0000000000
[    0.834548] mt7621-pci 1e140000.pcie: pcie0 no card, disable it (RST & CLK)
[    0.841443] mt7621-pci 1e140000.pcie: pcie2 no card, disable it (RST & CLK)
[    0.848383] mt7621-pci 1e140000.pcie: PCIE1 enabled
[    0.853215] PCI coherence region base: 0x60000000, mask/settings: 0xf0000002
[    0.860383] mt7621-pci 1e140000.pcie: PCI host bridge to bus 0000:00
...
```
When it is detected, the output looks like:
```
  root@OpenWrt:~# dmesg | grep PCIE
[    0.841804] mt7621-pci 1e140000.pcie: PCIE1 enabled
[    0.846654] mt7621-pci 1e140000.pcie: PCIE2 enabled
```

I believe I can work it around applying the old patch or similar, but it doesn't look like a general solution.
Please let me know which changes and improvements I can test on this device further.

DragonBluep commented:
@dfateyev It's strange. Your device doesn't look like a WE3526, but WE1326. The WE3526 uses pcie0 and pcie1, but your device uses pcie1 and pcie2. It seems that you have installed the wrong firmware? not sure.
https://forum.openwrt.org/t/mediatek-mt7613e-range-issue-on-zbt-we3526/91447/8

dfateyev commented:
> @dfateyev It's strange. Your device doesn't look like a WE3526, but WE1326. The WE3526 uses pcie0 and pcie1, but your device uses pcie1 and pcie2. It seems that you have installed the wrong firmware?

The model marked as WF3526-P by an OEM manufacturer, that's why I concluded it was WE3526. Although, these boards are pretty similar, referring to [techdata](https://openwrt.org/toh/hwdata/zbt/zbt_we3526) and [device page](https://openwrt.org/toh/zbtlink/zbt_we3526).
I can flash it as WE1326, but I think with the current version from repo we will face the same issues. Have we already figured out which changes are needed for this model exactly?

dfateyev commented:
> I can flash it as WE1326, but I think with the current version from repo we will face the same issues. Have we already figured out which changes are needed for this model exactly?

I have tried to flash it with WE1326 firmware from master, and (expectedly) got the same results with PCIE detection.
The dmesg output was the same, the only difference I got was both wireless interfaces were marked in Luci as "bgn".
Changing `PERST_DELAY_MS` to 500 gives a partial remedy, the same as by previous attempts.

I can experiment with this board more, but need to know what we're going to push upstream / which changes we need to test exactly? Whether we need DTS this PR, and what about `msleep` / `mdelay` usage?

wifiseguro commented:
@dfateyev 

Have you tried this patch? Just adding 75 ms mdelay or msleep right there fixes the issue, at least in my we1326
As your HW is the same o very similar, it might work for you too.

            --- a/drivers/staging/mt7621-pci/pci-mt7621.c
            +++ b/drivers/staging/mt7621-pci/pci-mt7621.c
            @@ -520,6 +520,7 @@
            list_del(&port->list);
            }
            }
            + mdelay(75);
            
            mt7621_pcie_reset_ep_deassert(pcie);

I have explained a few comments above https://github.com/openwrt/openwrt/pull/11220#issuecomment-1320951490

As stated by @paraka, my HW needs to wait a bit after making the phy_power-on of the physical port. I am trying to find out if that affects more devices as yours or the Netgear R6220

wifiseguro commented:
> > I can flash it as WE1326, but I think with the current version from repo we will face the same issues. Have we already figured out which changes are needed for this model exactly?
> 
> I have tried to flash it with WE1326 firmware from master, and (expectedly) got the same results with PCIE detection. The dmesg output was the same, the only difference I got was both wireless interfaces were marked in Luci as "bgn". Changing `PERST_DELAY_MS` to 500 gives a partial remedy, the same as by previous attempts.
> 
> I can experiment with this board more, but need to know what we're going to push upstream / which changes we need to test exactly? Whether we need DTS this PR, and what about `msleep` / `mdelay` usage?

Both you and me came to the same conclusions, the PERST_DELAY is not reliable for our boards. Browse my comments and you will realize that.
I am confident that this is the patch you need for your issue to be gone, as it was to me.  Please try it and once confirmed we can request to be pulled to master and 22.03 and wherever.
Regarding mdelay or msleep. Use msleep as the maintainers prefer that approach. They both works though, already tested.
Forget about DTS, expecially the pcie ports renaming part, it briks the device. Neither the GPIO part is what we need, at least to fix the cards detection. We need a PR. Please try it and inform back. I look forward your comments. Thanks

      --- a/drivers/staging/mt7621-pci/pci-mt7621.c
      +++ b/drivers/staging/mt7621-pci/pci-mt7621.c
      @@ -520,6 +520,7 @@
      list_del(&port->list);
      }
      }
      + mdelay(75);
      
      mt7621_pcie_reset_ep_deassert(pcie);

dfateyev commented:
@wifiseguro I think you are right, and to detect devices steadily both in master and 22.03 we need only a small delay before `mt7621_pcie_reset_ep_deassert()`. Though I have changed the patch slightly.

For master SNAPSHOT, in `patches-5.15`:
```
--- a/drivers/pci/controller/pcie-mt7621.c
+++ b/drivers/pci/controller/pcie-mt7621.c
@@ -374,6 +374,7 @@
 		}
 	}
 
+	msleep(PERST_DELAY_MS);
 	mt7621_pcie_reset_ep_deassert(pcie);
 
 	tmp = NULL;
```

For 22.03 SNAPSHOT, in `patches-5.10`:
```
--- a/drivers/staging/mt7621-pci/pci-mt7621.c
+++ b/drivers/staging/mt7621-pci/pci-mt7621.c
@@ -521,6 +521,7 @@
 		}
 	}
 
+	msleep(PERST_DELAY_MS);
 	mt7621_pcie_reset_ep_deassert(pcie);
 
 	tmp = NULL;
```
With these changes, I always get both PCIE cards detected, regardless of warm or cold restarts. It also shouldn't affect any other customers with more modern chips. We [need to use](https://www.kernel.org/doc/Documentation/timers/timers-howto.txt) `msleep()` there since we don't have atomic changes and the sleeping time is relatively large.

There are other issues discovered during testing:
- The 2.4Ghz interface (MT7603E) functionality is broken both in 22.03 and master. Either SMPS handling is still broken for this chip, or there is another issue. Reproduced on several WE3526 devices;
- The 5Ghz interface ( MT7612E) functionality is broken in master. The WiFi signal is very weak, TX power is set to 3 dBm and WiFi LED is not functional.

But these issues probably should be addressed separately. The chip detection issue seems to be solved.

DragonBluep commented:
> There are other issues discovered during testing:
> 
> * The 2.4Ghz interface (MT7603E) functionality is broken both in 22.03 and master. Either SMPS handling is still broken for this chip, or there is another issue. Reproduced on several WE3526 devices;
> * The 5Ghz interface ( MT7612E) functionality is broken in master. The WiFi signal is very weak, TX power is set to 3 dBm and WiFi LED is not functional;
> * WiFi interfaces MAC detection is broken in master, they are assigned with random MACs after each restart. There is an error in kernel logs about that.

I guess most of these problems only occur in kernel 5.15, right? If so, we still need to switch the pcie port.

dfateyev commented:
> I guess most of these problems only occur in kernel 5.15, right? If so, we still need to switch the pcie port.

The first issue is with mt76, this issue is long lasting. Other issues can be related to "kernel-5.15" or later changes in mt76.
I can test anything in master if there are any suggestions on what to test.

Also, I would appreciate some feedback from owners of this or similar boards on whether they have similar issues.

DragonBluep commented:
@dfateyev Theoretically, switching the pcie port (the first commit 63f5fdf61548eb3f0dde242361fd07d62fd3772d) can make the performance consistent with the 5.10 kernel. However, @wifiseguro reported that this would cause the system to fail to boot.

It will be good if you are willing to confirm this under 5.15 kernel (with your sleep time fix patch).
Related issues: https://github.com/openwrt/openwrt/pull/11284

DragonBluep commented:
> The model marked as WF3526-P by an OEM manufacturer, that's why I concluded it was WE3526. Although, these boards are pretty similar, referring to [techdata](https://openwrt.org/toh/hwdata/zbt/zbt_we3526) and [device page](https://openwrt.org/toh/zbtlink/zbt_we3526).
> [Wikidev](https://wikidevi.wi-cat.ru/ZBT_WE1326): "A Family Models Letter at the FCC claims that the following model names all use the same circuit and RF module: **ZBT-WE1326**, ZBT-WE1026-5G, **ZBT-WE3526**, ..."

@dfateyev In fact, their hardware is not the same. You need to flash the firmware for WE1326. As for 5GHz LED, you can add these properties to repair them.
```
		led {
			led-sources = <2>;
		};
```
I pushed some changes, so now you can just build and try the firmware for ZBT-WF3526-P. The generated binary name is still xxx-we1326-sysupgrade.bin

dfateyev commented:
> @dfateyev In fact, their hardware is not the same. You need to flash the firmware for WE1326. As for 5GHz LED, you can add these properties to repair them.
> 
> ```
> 		led {
> 			led-sources = <2>;
> 		};
> ```
> 
> WE1326 has a SD card slot, a MicroSD card slot and an internal pcie slot, all LEDs are green. WE3526 only has one MicroSD card slot, and one of the LEDs is red (two colors can be displayed). I pushed some changes, so now you can just build and try the firmware for ZBT-WF3526-P. The generated binary name is still xxx-we1326-sysupgrade.bin

@DragonBluep thanks for you efforts, I will try with WE1326 again, no problem.
Actually, the WiFi state LED on my device is 3-colored: it blinks green, orange and red, showing the appropriate WiFi interface activity (currently works fine on "22.03-SNAPSHOT").

DragonBluep commented:
> Actually, the WiFi state LED on my device is 3-colored: it blinks green, orange and red,

@dfateyev So your device should be a new design. 😂 But at least the pcie port is the same as WE1326. Maybe we can try to add a new device support after the wifi problem is solved.

dfateyev commented:
> > Actually, the WiFi state LED on my device is 3-colored: it blinks green, orange and red,
> 
> @dfateyev So your device should be a new design. joy But at least the pcie port is the same as WE1326. Maybe we can try to add a new device support after the wifi problem is solved.

I have applied all 3 commits from PR against [mt7621_zbtlink_zbt-we1326.dts](https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7621_zbtlink_zbt-we1326.dts) (adjusted b068e9d3758c80d9dd81f18749bdebce753e5103, it already cannot be applied after two others), built with [init delay](https://github.com/openwrt/openwrt/pull/11220#issuecomment-1331437590) patch, and flashed as WE1326.

Good news is that Wifi LED functionality restored; the invalid MAC errors in dmesg, 5GHz interface power and interface naming in LuCi are fixed with these changes. But the regressions with WiFi are still in place:
- 2.4Ghz interface (MT7603E) connections breaks after a short period of time under `iperf3` load;
- 5Ghz interface (MT7612E) shows speed in 2-3 times less than in 22.03-SNAPSHOT build.

@DragonBluep do we need GPIO reset part in PR if the device detection can be fixed with [init delay](https://github.com/openwrt/openwrt/pull/11220#issuecomment-1331437590) patch?

@paraka probably we can push [init delay](https://github.com/openwrt/openwrt/pull/11220#issuecomment-1331437590) patch into mainline instead of [previous changes](https://github.com/openwrt/openwrt/pull/11220#issuecomment-1320861537)? We also can provide them for 22.03 and master, in "patches-5.XX".

DragonBluep commented:
1. Do 2.4G and 5G LEDs work properly? They should be default on when WiFi is up and **keep off** when WiFi is down. It should blink when there is data transmission.
2. If it can work without `reset-gpio` change, then we can just forget it. it's not necessary.
3. `(MT7603E) connections breaks` should be a very very old topic in mt76, but iperf3 usually can't break it. The BT download will trigger this bug 100%. see https://github.com/openwrt/openwrt/issues/10074 You can try to reduce the tx power.
4. `5Ghz interface (MT7612E) shows speed in 2-3 times less than in 22.03-SNAPSHOT build.` I encountered this problem before. It seems to be related to nftable, but it should have been fixed about a month ago. Also, do not run iperf3 on the router. PC(iperf3 server) <--> Router LAN <--> WiFi client is the correct way to measure speed. Please make sure your Lan speed is 1 Gbps. 300+Mbps is a good performance for mt7612. The mt7615 is slightly better and can be close to 500 Mbps.

paraka commented:
@dfateyev I am waiting to see if any other changes are neccessary to submit all together to the mainline. For example to increase the PERS_DELAY_MS for netgear... Also, do you need to sleep 100 ms after making the pci phy on, or 75 ms is enough like it was for @wifiseguro ?

wifiseguro commented:
@DragonBluep 

		    led {
			    led-sources = <2>;
		    };

Ive been using this approach for  couple of years in my custom builds for the we1326 and would like to give my feedback

- There is a red led for 5ghz band and a green led for 2.4. Both Default on when bands are enabled at boot, and off when bands are disabled at boot. When both bands are on, the led shows as orange. When  there is 5ghz activity, you see it blink from orange to green (as the red is the one that is blinking), and with 2.4 activity you see it blink from orange to red (as the green blinks in this case)
- You can turn on the leds, but there is no way to turn them off once have been on, even you turn wifi off. 
- I.e , you boot the device with only  the 5 ghz band, you get just a red led that cannot be turn off. The same if you boot with both bands, you get an orange(red and green) led that cannot be turn off even you turn wifi off.

If you have any idea to workaround this I am willing to collaborate in the testing.

DragonBluep commented:
> So I restore the PERS_DELAY_MS to 100ms and I decide to dig a bit on my own patch, and I dont understand why, but the conclusion, also after many reboots with this patch, was that my device needs just 75ms delay!, but placed in a diferent place of the firmware code.

> I am putting just 75 ms delay right before function mt7621_pcie_reset_ep_deassert(pcie) and that fixes 2.4 pcie card detection,

> I think you are right, and to detect devices steadily both in master and 22.03 we need only a small delay before `mt7621_pcie_reset_ep_deassert()`.

**So, let me make a simple summary. For the ZBT WE1326/WF3526-P, we just need to sleep 75ms (maybe 100ms for other devices?) before `mt7621_pcie_reset_ep_deassert(pcie);`, and don't need to change the `PERST_DELAY_MS`**

The Netgear series devices may also not need to change the `PERS_DELAY_MS`, it seems that in the original patch delay before 
 `mt7621_pcie_reset_ep_deassert(pcie);` is enough. In any case, this problem can never be solved without test and feedback.

@wifiseguro I guess the LEDs should work now.
@dfateyev I want to confirm that your model is Zbtlink ZBT-W**F**3526-P, right? I've added it.

dfateyev commented:
> 1. Do 2.4G and 5G LEDs work properly? They should be default on when WiFi is up and **keep off** when WiFi is down. It should blink when there is data transmission.
...
> <@wifiseguro>  There is a red led for 5ghz band and a green led for 2.4. Both Default on when bands are enabled at boot, and off when bands are disabled at boot. When both bands are on, the led shows as orange. When there is 5ghz activity, you see it blink from orange to green (as the red is the one that is blinking), and with 2.4 activity you see it blink from orange to red (as the green blinks in this case)
...

It's exactly as on my device, with one difference: WiFi LED can be properly turned off once, right after the device boot. Having two radio interfaces enabled, I can turn off 5GHz interface (LED starts blinking green), then turn off 2.4Ghz interface (LED turns off). I can enable interfaces back, LED starts blinking different colors again, but switching the interfaces off again doesn't make LED switching off.
I suspect the reason is that radio actually is not switching off completely: being disabled, both interfaces are still exposed in the air with their SSID, but connections to AP fail.

> 2. If it can work without `reset-gpio` change, then we can just forget it. it's not necessary.

I think the GPIO part is redundant: I have applied only fe23a1b815841f8fb61c43f8135f1922484a7cf7 and e51cd7d4dc0578412b06816422fbac28ba22b2ee, and PCIE detection works as before.

> 3. `(MT7603E) connections breaks` should be a very very old topic in mt76, but iperf3 usually can't break it. The BT download will trigger this bug 100%. see [ramips: MT7628: WiFi signal disappear under heavy load #10074](https://github.com/openwrt/openwrt/issues/10074) You can try to reduce the tx power.

Unfortunately, it didn't help: I have decreased the power to 17 dBm (50%), but still experience WiFi loss without restoring, after 10-30sec of WiFi channel intensive usage. The 2.4Ghz radio is completely unusable as for now.

> 4. `5Ghz interface (MT7612E) shows speed in 2-3 times less than in 22.03-SNAPSHOT build.` I encountered this problem before. It seems to be related to nftable, but it should have been fixed about a month ago. Also, do not run iperf3 on the router. PC(iperf3 server) <--> Router LAN <--> WiFi client is the correct way to measure speed. Please make sure your Lan speed is 1 Gbps. 300+Mbps is a good performance for mt7612. The mt7615 is slightly better and can be close to 500 Mbps.

I believe the recent master helped a bit, they provided newer `mac80211`. I still experience throughput degrade compared to 22.03-SNAPSHOT build, now it's ~1.5 times less, so there is some improvement.

> @dfateyev I want to confirm that your model is Zbtlink ZBT-WF3526-P, right? I've added it.

It's not Zbtlink-branded, probably a ZBT's sub-brand or an OEM vendor. The full name of this device: Wiflyer WF3526-P.
Amazon provides a bunch of Wiflyer devices, all of them look like re-branded Zbtlink ones.
I'm not even sure whether we should add it: previously it worked fine with WE3526 firmware on older OpenWrt, but if you feel it would be better to add it to the list, it's okay.

> So, let me make a simple summary. For the ZBT WE1326/WF3526-P, we just need to sleep 75ms (maybe 100ms for other devices?) before mt7621_pcie_reset_ep_deassert(pcie);, and don't need to change the PERST_DELAY_MS

I think yes. Changing `PERST_DELAY_MS` didn't made a trick for me, but only increased initialization time w/o stable effect.
I prefer to use 100ms to be compliant with other `msleep()` calls there, thus these delays can be regulated by `PERST_DELAY_MS` value, and not by a hard-coded value which we introduce.

> The Netgear series devices may also not need to change the PERS_DELAY_MS, it seems that in the original patch delay before mt7621_pcie_reset_ep_deassert(pcie); is enough. In any case, this problem can never be solved without test and feedback.

Probably, yes — we need more feedback from Netgear users, as I don't have any Netgear device.

wifiseguro commented:
@DragonBluep  
I confirm that now boths leds can be disabled but as @dfateyev pointed out the radio might not being disabled completely (I can find the SSI available on air even Luci says it's disabled) And I found this workaround for that: pressing  Luci's  "reset" button of the band you want to disable prior to disable it. Doing that the radio is completelly disabled, as it's led is. 

About the commit for correcting the Pcie numbering, I confirm that is necessary for we1326 in kernel 5.15, or else it doesnt detect 5ghz band. On kernel 5.10 it makes the device unresponding on boot. A few messages above, you where asking me to test  that commit and I told you I couldnt boot with it applied. The thing is that I was building on 5.10. Could it be that you were expecting me to test on 5.15? 
Anyway I confirm that correct pcie port numering is needed on 5.15 to fix 5ghz band,  and it is not for 5.10.

DragonBluep commented:
@wifiseguro Yes, PCIe port number fixes are only applicable to the upcoming 5.15 kernel, as previously mentioned. https://github.com/openwrt/openwrt/pull/11220#issuecomment-1321403626
Now let's wait for the mainline switch to the kernel 5.15, and then we can reopen the PR.

dfateyev commented:
> 3. `(MT7603E) connections breaks` should be a very very old topic in mt76, but iperf3 usually can't break it. The BT download will trigger this bug 100%. see [ramips: MT7628: WiFi signal disappear under heavy load #10074](https://github.com/openwrt/openwrt/issues/10074) You can try to reduce the tx power.

I'm mostly concerned about 2.4Ghz interface stability under "kernel-5.15". Your suggestions on reducing the transmit power led me to more experiments, and with 25% of max power (14 dBm / 25mW) I get a connection that lives up longer, but eventually fails within a couple of minutes. I didn't require to do that with 5.10 or earlier, so consider it a regression.

>  The full name of this device: Wiflyer WF3526-P

@DragonBluep if you're going to add this device, have you already adjusted the name?

DragonBluep commented:
@dfateyev The name should be OK now. I can't help more about mt7603, you can try to give feedback to developers in [mt76](https://github.com/openwrt/mt76).

paraka commented:
I have already sent proposed changes to mainline:
https://lore.kernel.org/linux-pci/20221209071703.2891714-1-sergio.paracuellos@gmail.com/T/#u

dfateyev commented on target/linux/ramips/dts/mt7621_zbtlink_zbt-we1326.dts:
```diff
@@ -127,6 +127,11 @@
 		reg = <0x0000 0 0 0 0>;
 		mediatek,mtd-eeprom = <&factory 0x8000>;
 		ieee80211-freq-limit = <5000000 6000000>;
+		
```
Small nit-pick: please clean whitespaces in empty lines, in both patch hunks, before merging

dfateyev commented:
We also need a similar patch for 21.02 branch, to ensure old devices to be initialized properly (otherwise, there are the same detection issues). 
For 21.02 SNAPSHOT, in `patches-5.04`:
```
--- a/drivers/staging/mt7621-pci/pci-mt7621.c
+++ b/drivers/staging/mt7621-pci/pci-mt7621.c
@@ -521,6 +521,7 @@
 		}
 	}
 
+	mdelay(PERST_DELAY_MS);
 	mt7621_pcie_reset_ep_deassert(pcie);
 
 	tmp = NULL;
```
Used `mdelay()` here since in older mt76 the internal delays were regulated with mdelay.
These changes tested with the same devices, also the suggestions from [upstream](https://github.com/openwrt/openwrt/pull/11220#issuecomment-1344028209) on macros naming can be applied here.

dfateyev commented:
> I have already sent proposed changes to mainline: https://lore.kernel.org/linux-pci/20221209071703.2891714-1-sergio.paracuellos@gmail.com/T/#u

@paraka is there any news on adding this patch to the upstream?
can we add it to openwrt 21.01 (5.04), 22.03 (5.10) and master (5.15)?

paraka commented:
> > I have already sent proposed changes to mainline: https://lore.kernel.org/linux-pci/20221209071703.2891714-1-sergio.paracuellos@gmail.com/T/#u
> 
> @paraka is there any news on adding this patch to the upstream? can we add it to openwrt 21.02 (5.4), 22.03 (5.10) and master (5.15)?

I sent a ping to the PCI tree maintainers a few days ago but still no response:
https://lore.kernel.org/linux-pci/CAMhs-H8sSvn5bfHde-RT0ejqpCujCWvgRFJLtbLXqPcnvGdgtQ@mail.gmail.com/T/#t

DragonBluep commented:
@dfateyev I backport the patch for the 5.4/5.10/5.15 kernel. You can try to `cherry-pick` it to the corresponding branches and test it. If you have time, you can PR them to branches  21.02 (5.4), 22.03 (5.10) and master (5.15). The 5.15 kernel is still under test. We can apply this fix in advance.

dfateyev commented:
> I backport the patch for the 5.4/5.10/5.15 kernel. You can try to `cherry-pick` it to the corresponding branches and test it. If you have time, you can PR them to branches 21.02 (5.4), 22.03 (5.10) and master (5.15). The 5.15 kernel is still under test. We can apply this fix in advance.

@DragonBluep thanks for the backport, it looks good. The only change is needed: please use `mdelay()` for openwrt 21.02 (patches-5.4) as [pointed above](https://github.com/openwrt/openwrt/pull/11220#issuecomment-1368435676). The mt76 version in older branches uses `mdelay()` to regulate internal delays.

DragonBluep commented:
@dfateyev Since this series of patches are backported from the mailing lists, I think they should use the same function. delay() is usually more accurate than sleep(), however it will consume unnecessary CPU resources.

dfateyev commented:
> Since this series of patches are backported from the mailing lists, I think they should use the same function. delay() is usually more accurate than sleep(), however it will consume unnecessary CPU resources.

@DragonBluep my concern is that in older staging mt76 there is no single `sleep()` in the `"pci-mt7621.c"`, but multiple `delay()` present. We probably need to follow the same logic, not to mix both approaches. Also, the 21.02 patch with `mdelay()` is already tested on two MT7621 devices for a while.
As for newer Openwrt versions, there is undoubtedly `msleep()` should be used there.

DragonBluep commented:
Update:

1. Remove the untested PCIe port changes of the NETGEAR series devices. If NETGEAR users encounter the same problem in the future, the information in this PR should be enough to help them fix it. https://github.com/openwrt/openwrt/commit/29447ea2c58a423031ec4ed827bffcd574596ade
2. Remove `Delay phy ports initialization` patch since the final upstreamed version was sent to OpenWrt Mailing List some days ago. https://patchwork.ozlabs.org/project/openwrt/patch/mailman.13941.1675951439.3064139.openwrt-devel@lists.openwrt.org/ (This is my first time using Mailing List. There seems to be some format errors in the email. If necessary, I'll try to submit a v2 version.)

DragonBluep committed:
ramips: add alternative device name for Wiflyer WF3526-P

Wiflyer WF3526-P and Zbtlink ZBT-WE1326 have the same circuit design.
Installing the misunderstading firmware of ZBT-WE3526 will cause Wi-Fi
not work due to allocate the wrong pcie port. Add alternative name to
help users easily build or download the correct firmware.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: correct the PCIe port number for Zbtlink ZBT-WE1326

MT7621 gets a new PCIe driver in the 5.15+ kernel. Allocating wrong PCIe
port will cause the PCIe NIC to not work properly. This commit fixes
the wrong port numbers on Zbtlink ZBT-WE1326.

According to the bootlog, MT7612E (5 GHz) is connected to pcie1, and
MT7603E (2 GHz) is connected to pcie2:
[4.197658] mt7621-pci 1e140000.pcie: pcie0 no card, disable it (RST & CLK)
[4.204609] mt7621-pci 1e140000.pcie: PCIE1 enabled
[4.209476] mt7621-pci 1e140000.pcie: PCIE2 enabled
...
[4.307988] pci 0000:01:00.0: [14c3:7662] type 00 class 0x028000
[4.367206] pci 0000:02:00.0: [14c3:7603] type 00 class 0x028000

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep commented:
@paraka A feedback: MT7915 datasheet said PERST_N need to keep low for at least 500 ms to reset the chip. I think this is the reason why additional delay is needed (MT7603 and MT7612 should also require the same time). 
https://edit.wpgdadawant.com/uploads/news_file/blog/2020/2406/tech_files/blog_2406_suggest_other_file.pdf
